File: languages/java/src/main/java/com/bloock/sdk/client/RecordBuilder.java
package com.bloock.sdk.client;

import com.bloock.sdk.bridge.Bridge;
import com.bloock.sdk.bridge.proto.BloockAuthenticityEntities;
import com.bloock.sdk.bridge.proto.BloockAvailabilityEntities;
import com.bloock.sdk.bridge.proto.BloockConfig.ConfigData;
import com.bloock.sdk.bridge.proto.BloockEncryptionEntities;
import com.bloock.sdk.bridge.proto.BloockRecord.*;
import com.bloock.sdk.bridge.proto.BloockRecordEntities;
import com.bloock.sdk.bridge.proto.BloockRecordEntities.RecordTypes;
import com.bloock.sdk.bridge.proto.BloockShared.Error;
import com.bloock.sdk.entity.authenticity.Signer;
import com.bloock.sdk.entity.encryption.Encrypter;
import com.bloock.sdk.entity.record.Record;
import com.bloock.sdk.entity.record.RecordDetails;
import com.google.protobuf.ByteString;

/**
 * Assists in constructing records with various configurations.
 */
public class RecordBuilder {
  Object payload;
  RecordTypes type;
  ConfigData configData;
  BloockAuthenticityEntities.Signer signer;
  BloockEncryptionEntities.Encrypter encrypter;
  BloockEncryptionEntities.Encrypter decrypter;

  /**
   * Creates a new RecordBuilder with default configuration.
   * 
   * @param payload
   * @param type
   * @param configData
   */
  RecordBuilder(Object payload, RecordTypes type, ConfigData configData) {
    this.payload = payload;
    this.type = type;
    this.configData = configData;
  }

  /**
   * Sets the signer for the RecordBuilder.
   * 
   * @param signer
   * @return
   */
  public RecordBuilder withSigner(Signer signer) {
    this.signer = signer.toProto();
    return this;
  }

  /**
   * Sets the encrypter for the RecordBuilder.
   * 
   * @param encrypter
   * @return
   */
  public RecordBuilder withEncrypter(Encrypter encrypter) {
    this.encrypter = encrypter.toProto();
    return this;
  }

  /**
   * Sets the decrypter for the RecordBuilder.
   * 
   * @param decrypter
   * @return
   */
  public RecordBuilder withDecrypter(Encrypter decrypter) {
    this.decrypter = decrypter.toProto();
    return this;
  }

  /**
   * Constructs a record based on the RecordBuilder's configuration.
   * 
   * @return
   * @throws Exception
   */
  public Record build() throws Exception {
    Bridge bridge = new Bridge();

    RecordBuilderResponse response;

    switch (this.type) {
      case BYTES: {
        RecordBuilderFromBytesRequest.Builder builder = RecordBuilderFromBytesRequest.newBuilder()
            .setConfigData(this.configData)
            .setPayload(ByteString.copyFrom((byte[]) this.payload));

        if (this.signer != null) {
          builder.setSigner(this.signer);
        }
        if (this.encrypter != null) {
          builder.setEncrypter(this.encrypter);
        }
        if (this.decrypter != null) {
          builder.setDecrypter(this.decrypter);
        }

        response = bridge.getRecord().buildRecordFromBytes(builder.build());
        break;
      }
      case FILE: {
        RecordBuilderFromFileRequest.Builder builder = RecordBuilderFromFileRequest.newBuilder()
            .setConfigData(this.configData)
            .setPayload(ByteString.copyFrom((byte[]) this.payload));

        if (this.signer != null) {
          builder.setSigner(this.signer);
        }
        if (this.encrypter != null) {
          builder.setEncrypter(this.encrypter);
        }
        if (this.decrypter != null) {
          builder.setDecrypter(this.decrypter);
        }

        response = bridge.getRecord().buildRecordFromFile(builder.build());
        break;
      }
      case HEX: {
        RecordBuilderFromHexRequest.Builder builder = RecordBuilderFromHexRequest.newBuilder()
            .setConfigData(this.configData)
            .setPayload((String) this.payload);

        if (this.signer != null) {
          builder.setSigner(this.signer);
        }
        if (this.encrypter != null) {
          builder.setEncrypter(this.encrypter);
        }
        if (this.decrypter != null) {
          builder.setDecrypter(this.decrypter);
        }

        response = bridge.getRecord().buildRecordFromHex(builder.build());
        break;
      }
      case JSON: {
        RecordBuilderFromJSONRequest.Builder builder = RecordBuilderFromJSONRequest.newBuilder()
            .setConfigData(this.configData)
            .setPayload((String) this.payload);

        if (this.signer != null) {
          builder.setSigner(this.signer);
        }
        if (this.encrypter != null) {
          builder.setEncrypter(this.encrypter);
        }
        if (this.decrypter != null) {
          builder.setDecrypter(this.decrypter);
        }

        response = bridge.getRecord().buildRecordFromJson(builder.build());
        break;
      }
      case RECORD: {
        RecordBuilderFromRecordRequest.Builder builder = RecordBuilderFromRecordRequest.newBuilder()
            .setConfigData(this.configData)
            .setPayload((BloockRecordEntities.Record) this.payload);

        if (this.signer != null) {
          builder.setSigner(this.signer);
        }
        if (this.encrypter != null) {
          builder.setEncrypter(this.encrypter);
        }
        if (this.decrypter != null) {
          builder.setDecrypter(this.decrypter);
        }

        response = bridge.getRecord().buildRecordFromRecord(builder.build());
        break;
      }
      case STRING: {
        RecordBuilderFromStringRequest.Builder builder = RecordBuilderFromStringRequest.newBuilder()
            .setConfigData(this.configData)
            .setPayload((String) this.payload);

        if (this.signer != null) {
          builder.setSigner(this.signer);
        }
        if (this.encrypter != null) {
          builder.setEncrypter(this.encrypter);
        }
        if (this.decrypter != null) {
          builder.setDecrypter(this.decrypter);
        }

        response = bridge.getRecord().buildRecordFromString(builder.build());
        break;
      }
      case LOADER: {
        RecordBuilderFromLoaderRequest.Builder builder = RecordBuilderFromLoaderRequest.newBuilder()
            .setConfigData(this.configData)
            .setLoader((BloockAvailabilityEntities.Loader) this.payload);

        if (this.signer != null) {
          builder.setSigner(this.signer);
        }
        if (this.encrypter != null) {
          builder.setEncrypter(this.encrypter);
        }
        if (this.decrypter != null) {
          builder.setDecrypter(this.decrypter);
        }

        response = bridge.getRecord().buildRecordFromLoader(builder.build());
        break;
      }
      default:
        throw new Exception("Invalid type");
    }

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return Record.fromProto(response.getRecord(), this.configData);
  }

  /**
   * Gets details about other Bloock services (Integrity, Authenticity,
   * Encryption, Availability) configured in the RecordBuilder.
   * 
   * @return
   * @throws Exception
   */
  public RecordDetails getDetails() throws Exception {
    Bridge bridge = new Bridge();

    GetDetailsRequest.Builder builder = GetDetailsRequest.newBuilder()
        .setConfigData(this.configData)
        .setPayload(ByteString.copyFrom((byte[]) this.payload));

    GetDetailsResponse response = bridge.getRecord().getDetails(builder.build());

    if (response.getError() != Error.getDefaultInstance()) {
      throw new Exception(response.getError().getMessage());
    }

    return RecordDetails.fromProto(response.getDetails());
  }
}

File: languages/php/Bloock/Client/RecordBuilder.php
<?php

namespace Bloock\Client;

use Bloock\Bridge\Bridge;
use Bloock\ConfigData;
use Bloock\Entity\Authenticity\Signer;
use Bloock\Entity\Encryption\Encrypter;
use Bloock\Entity\Record\Record;
use Bloock\Entity\Record\RecordDetails;
use Bloock\GetDetailsRequest;
use Bloock\RecordBuilderFromBytesRequest;
use Bloock\RecordBuilderFromFileRequest;
use Bloock\RecordBuilderFromHexRequest;
use Bloock\RecordBuilderFromJSONRequest;
use Bloock\RecordBuilderFromLoaderRequest;
use Bloock\RecordBuilderFromRecordRequest;
use Bloock\RecordBuilderFromStringRequest;
use Bloock\RecordTypes;
use Exception;

class RecordBuilder
{
    private $payload;
    private int $type;
    private ConfigData $configData;
    private ?\Bloock\Signer $signer = null;
    private ?\Bloock\Encrypter $encrypter = null;
    private ?\Bloock\Encrypter $decrypter = null;

    public function __construct($payload, int $recordTypes, ConfigData $configData)
    {
        $this->payload = $payload;
        $this->type = $recordTypes;
        $this->configData = $configData;
    }

    public function withSigner(Signer $signer): RecordBuilder
    {
        $this->signer = $signer->toProto();
        return $this;
    }

    public function withEncrypter(Encrypter $encrypter): RecordBuilder
    {
        $this->encrypter = $encrypter->toProto();
        return $this;
    }

    public function withDecrypter(Encrypter $decrypter): RecordBuilder
    {
        $this->decrypter = $decrypter->toProto();
        return $this;
    }

    public function build(): Record
    {
        $bridge = new Bridge();

        $res = null;
        switch ($this->type) {
            case RecordTypes::BYTES:
                $req = new RecordBuilderFromBytesRequest();
                $req->setConfigData($this->configData)->setPayload(implode(array_map("chr", $this->payload)));

                if ($this->signer != null) {
                    $req->setSigner($this->signer);
                }

                if ($this->encrypter != null) {
                    $req->setEncrypter($this->encrypter);
                }

                if ($this->decrypter != null) {
                    $req->setDecrypter($this->decrypter);
                }

                $res = $bridge->record->BuildRecordFromBytes($req);
                break;
            case RecordTypes::FILE:
                $req = new RecordBuilderFromFileRequest();
                $req->setConfigData($this->configData)->setPayload(implode(array_map("chr", $this->payload)));

                if ($this->signer != null) {
                    $req->setSigner($this->signer);
                }

                if ($this->encrypter != null) {
                    $req->setEncrypter($this->encrypter);
                }

                if ($this->decrypter != null) {
                    $req->setDecrypter($this->decrypter);
                }

                $res = $bridge->record->BuildRecordFromFile($req);
                break;
            case RecordTypes::HEX:
                $req = new RecordBuilderFromHexRequest();
                $req->setConfigData($this->configData)->setPayload($this->payload);

                if ($this->signer != null) {
                    $req->setSigner($this->signer);
                }

                if ($this->encrypter != null) {
                    $req->setEncrypter($this->encrypter);
                }

                if ($this->decrypter != null) {
                    $req->setDecrypter($this->decrypter);
                }

                $res = $bridge->record->BuildRecordFromHex($req);
                break;
            case RecordTypes::JSON:
                $req = new RecordBuilderFromJSONRequest();
                $req->setConfigData($this->configData)->setPayload($this->payload);

                if ($this->signer != null) {
                    $req->setSigner($this->signer);
                }

                if ($this->encrypter != null) {
                    $req->setEncrypter($this->encrypter);
                }

                if ($this->decrypter != null) {
                    $req->setDecrypter($this->decrypter);
                }

                $res = $bridge->record->BuildRecordFromJson($req);
                break;
            case RecordTypes::RECORD:
                $req = new RecordBuilderFromRecordRequest();
                $req->setConfigData($this->configData)->setPayload($this->payload);

                if ($this->signer != null) {
                    $req->setSigner($this->signer);
                }

                if ($this->encrypter != null) {
                    $req->setEncrypter($this->encrypter);
                }

                if ($this->decrypter != null) {
                    $req->setDecrypter($this->decrypter);
                }

                $res = $bridge->record->BuildRecordFromRecord($req);
                break;
            case RecordTypes::STRING:
                $req = new RecordBuilderFromStringRequest();
                $req->setConfigData($this->configData)->setPayload($this->payload);

                if ($this->signer != null) {
                    $req->setSigner($this->signer);
                }

                if ($this->encrypter != null) {
                    $req->setEncrypter($this->encrypter);
                }

                if ($this->decrypter != null) {
                    $req->setDecrypter($this->decrypter);
                }

                $res = $bridge->record->BuildRecordFromString($req);
                break;
            case RecordTypes::LOADER:
                $req = new RecordBuilderFromLoaderRequest();
                $req->setLoader($this->payload)->setConfigData($this->configData);

                if ($this->signer != null) {
                    $req->setSigner($this->signer);
                }

                if ($this->encrypter != null) {
                    $req->setEncrypter($this->encrypter);
                }

                if ($this->decrypter != null) {
                    $req->setDecrypter($this->decrypter);
                }

                $res = $bridge->record->BuildRecordFromLoader($req);
                break;
            default:
                throw new Exception("Invalid type");
        }

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return Record::fromProto($res->getRecord(), $this->configData);
    }

    public function getDetails(): RecordDetails
    {
        $bridge = new Bridge();

        $req = new GetDetailsRequest();
        $req->setPayload(implode(array_map("chr", $this->payload)))->setConfigData($this->configData);

        $res = $bridge->record->GetDetails($req);

        if ($res->getError() != null) {
            throw new Exception($res->getError()->getMessage());
        }

        return RecordDetails::fromProto($res->getDetails());
    }
}
